File: crates/taflow-core/src/indicators/day_of_week_return_profile.rs
use crate::error::TaResult;

#[derive(Debug, Clone, Copy, PartialEq)]
/// Mean close returns ordered from Monday through Sunday.
pub struct DayOfWeekReturnProfileValue {
    pub bins: [f64; 7],
}

/// Running mean close return for Monday through Sunday.
#[derive(Debug, Clone)]
pub struct DayOfWeekReturnProfile {
    offset_minutes: i32,
    previous_close: Option<f64>,
    sums: [f64; 7],
    counts: [u64; 7],
    value: Option<DayOfWeekReturnProfileValue>,
}

impl DayOfWeekReturnProfile {
    /// Create an empty weekday profile with a fixed UTC offset.
    pub fn new(utc_offset_minutes: i32) -> TaResult<Self> {
        Ok(Self {
            offset_minutes: utc_offset_minutes,
            previous_close: None,
            sums: [0.0; 7],
            counts: [0; 7],
            value: None,
        })
    }

    /// Append one OHLCV bar and Unix-nanosecond timestamp.
    pub fn append(
        &mut self,
        _open: f64,
        _high: f64,
        _low: f64,
        close: f64,
        _volume: f64,
        timestamp: i64,
    ) -> Option<DayOfWeekReturnProfileValue> {
        let Some(previous) = self.previous_close.replace(close) else {
            return None;
        };
        let local_seconds =
            timestamp.div_euclid(1_000_000_000) + i64::from(self.offset_minutes) * 60;
        let weekday = (local_seconds.div_euclid(86_400) + 3).rem_euclid(7) as usize;
        self.sums[weekday] += if previous == 0.0 {
            0.0
        } else {
            close / previous - 1.0
        };
        self.counts[weekday] += 1;
        let mut bins = [0.0; 7];
        for index in 0..7 {
            if self.counts[index] > 0 {
                bins[index] = self.sums[index] / self.counts[index] as f64;
            }
        }
        self.value = Some(DayOfWeekReturnProfileValue { bins });
        self.value
    }

    /// Return Monday-to-Sunday means after the first return exists.
    pub fn value(&self) -> Option<DayOfWeekReturnProfileValue> {
        self.value
    }
    /// Restore fresh-state behavior and clear all weekday accumulators.
    pub fn reset(&mut self) {
        self.previous_close = None;
        self.sums = [0.0; 7];
        self.counts = [0; 7];
        self.value = None;
    }
}

File: python/taflow/indicators/day_of_week_return_profile.py
"""Day-of-week return profile adapter."""

from typing import Any

import numpy as np

from .._native import DayOfWeekReturnProfile as _Native
from .._series import as_float64_series, as_int64_series


class DayOfWeekReturnProfile:
    """Compute expanding mean close returns for Monday through Sunday.

    Returns are assigned to the local weekday of the newer bar. Rust owns all
    arithmetic and returns a ``(bars, 7)`` history ordered Monday through
    Sunday; the first row is ``NaN``. This maps to Wickra ``DayOfWeekProfile``.

    Args:
        open: Required opening-price history.
        high: Required high-price history.
        low: Required low-price history.
        close: Required closing-price history.
        volume: Required volume history for the common OHLCV contract.
        timestamp: Chronological Unix timestamps in nanoseconds.
        utc_offset_minutes: Fixed offset applied before weekday selection.

    Raises:
        ValueError: If the input histories have different lengths.
    """

    def __init__(
        self,
        utc_offset_minutes: int = 0,
    ) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native(utc_offset_minutes)

    def append(
        self,
        open: float,
        high: float,
        low: float,
        close: float,
        volume: float,
        timestamp: int,
    ) -> "DayOfWeekReturnProfile":
        """Append one OHLCV bar and Unix-nanosecond timestamp; return ``self``."""
        self._state.append(
            float(open), float(high), float(low), float(close),
            float(volume), int(timestamp),
        )
        return self

    def extend(
        self,
        open: Any,
        high: Any,
        low: Any,
        close: Any,
        volume: Any,
        timestamp: Any,
    ) -> "DayOfWeekReturnProfile":
        """Append aligned OHLCV and timestamp histories and return ``self``."""
        arrays = tuple(
            as_float64_series(item) for item in (open, high, low, close, volume)
        ) + (as_int64_series(timestamp),)
        if len({len(item) for item in arrays}) != 1:
            raise ValueError("OHLCV and timestamp inputs must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> list[float] | None:
        """Return the seven current weekday means, or ``None`` during warm-up."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned native output with Monday-to-Sunday columns."""
        return self._state.compute()

    def reset(self) -> "DayOfWeekReturnProfile":
        """Clear native state and history and return ``self``."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)
